libcache/cache.c

```diff
@@ -0,0 +1,650 @@
+{
+	cacheline_t *linePtr = NULL;
+
+	linePtr = cache_updateTimestamps(set, tag);
+
+	return (linePtr != NULL) ? (unsigned char *)(linePtr)->data + offset : NULL;
+}
+
+
+static uint64_t cache_compOffset(const cachectx_t *cache, const uint64_t addr)
+{
+	return addr & cache->offMask;
+}
+
+
+static uint64_t cache_compSet(const cachectx_t *cache, const uint64_t addr)
+{
+	return (addr >> cache->offBitsNum) & cache->setMask;
+}
+
+
+static uint64_t cache_compTag(const cachectx_t *cache, const uint64_t addr)
+{
+	return (addr >> (cache->offBitsNum + cache->setBitsNum)) & cache->tagMask;
+}
+
+
+/* TODO: add error handling */
+ssize_t cache_write(cachectx_t *cache, uint64_t addr, void *buffer, size_t count, int policy)
+{
+	ssize_t position = 0, ret = 0;
+	void *data = NULL, *dest = NULL;
+	unsigned char flags = 0;
+	uint64_t index = 0, offset = 0, tag = 0;
+	size_t pieceSize, left = 0, remainder = 0;
+	cacheline_t line, *linePtr;
+
+	if (cache == NULL || buffer == NULL) {
+		return -1;
+	}
+
+	if (count == 0 || (policy != LIBCACHE_WRITE_BACK && policy != LIBCACHE_WRITE_THROUGH)) {
+		return 0;
+	}
+
+	offset = cache_compOffset(cache, addr);
+
+	left = count;
+
+	while (1) {
+		index = cache_compSet(cache, addr);
+		tag = cache_compTag(cache, addr);
+
+		if (left == count) {
+			if (count <= (cache->lineSize - offset)) {
+				pieceSize = count;
+			}
+			else {
+				pieceSize = cache->lineSize - offset;
+				remainder = (left - pieceSize) % cache->lineSize;
+			}
+			addr -= offset;
+		}
+		else if (left == remainder) {
+			pieceSize = remainder;
+			offset = 0;
+		}
+		else {
+			pieceSize = cache->lineSize;
+			offset = 0;
+		}
+
+		data = cache_readFromSet(&cache->sets[index], tag, offset);
+
+		if (data == NULL) {
+			data = calloc(cache->lineSize, sizeof(unsigned char));
+			dest = (unsigned char *)data + offset;
+			memcpy(dest, buffer + position, pieceSize);
+			SET_VALID(flags);
+
+			line.tag = tag;
+			line.data = data;
+
+			line.flags = flags;
+			line.offset = offset;
+
+			position += cache_writeToSet(cache->writeCb, &cache->sets[index], &line, addr, cache->lineSize, -1);
+		}
+		else {
+			linePtr = cache_findLine(&cache->sets[index], tag);
+
+			ret = cache_flushLine(cache->writeCb, linePtr, addr, cache->lineSize);
+			if (ret == -1) {
+				position = -1;
+				break;
+			}
+
+			dest = (unsigned char *)linePtr->data + offset;
+			memcpy(dest, buffer + position, pieceSize);
+
+			ret = cache_executePolicy(cache->writeCb, linePtr, addr, cache->lineSize, policy);
+
+			position += pieceSize;
+		}
+
+		left -= pieceSize;
+
+		if (left == 0) {
+			break;
+		}
+
+		addr += cache->lineSize;
+	}
+
+	return position;
+}
+
+
+static ssize_t cache_readOnMiss(cachectx_t *cache, const uint64_t addr, const uint64_t offset, void *buffer, size_t position, size_t count)
+{
+	ssize_t read = 0, written = 0;
+	void *temp = NULL;
+
+	temp = calloc(cache->lineSize, sizeof(unsigned char));
```

Comment:
This is very inefficient approach - you allocate lineSize buffer to read data, then in cache_write you again allocate lineSize buffer and copy the data between buffer and in the end you free first buffer. Cut the middle man